shared/src/main/scala/datasource.scala

```diff
@@ -25,21 +25,26 @@ import cats.syntax.all._
 
 /**
  * A `DataSource` is the recipe for fetching a certain identity `I`, which yields
- * results of type `A`.
+ * results of type `A` performing an effect of type `F[_]`.
  */
-trait DataSource[I, A] {
+trait DataSource[F[_], I, A] {
   /** The name of the data source.
    */
   def name: String
 
+  /** The identity of the data source.
+   */
+  def identity: DataSourceIdentity =
+    DataSourceIdentity(name)
+
   /** Fetch one identity, returning a None if it wasn't found.
    */
-  def fetch[F[_] : ConcurrentEffect](id: I): F[Option[A]]
+  def fetch(id: I)(implicit C: ConcurrentEffect[F]): F[Option[A]]
 
   /** Fetch many identities, returning a mapping from identities to results. If an
    * identity wasn't found, it won't appear in the keys.
    */
-  def batch[F[_] : ConcurrentEffect](ids: NonEmptyList[I]): F[Map[I, A]] =
+  def batch(ids: NonEmptyList[I])(implicit C: ConcurrentEffect[F]): F[Map[I, A]] =
```

Comment:
So, one problem with using `implicit` parameters is that you can not override them or remove them from a method in the subclass: the subclasses must keep a similar signature to the parent. 

We could leave here no implementation, or add one that only used an `Applicative` constraint (to use with `List.traverse`), and then move the implementation using `ConcurrentEffect` to a different mix-in trait. 

```Scala
trait ConcurrentBatch[F[_]] extends DataSource[F] {
   implicit protected ConcF: ConcurrentEffect[F] 
   def batch(ids: NonEmptyList[I]): F[Map[I, A]] = /****/
}
```

Reply:
The `ConcurrentEffect` type class is something we can't escape from since it'll be required by Fetch when running or creating fetches. Having the implicit `ConcurrentEffect` in scope is something that you want when defining data fetches, since very few of those will require only `Applicative`. 

Remember that Fetch is a library for reordering and optimizing IO, in particular reading data. In practice, you will be using the `ConcurrentEffect` API to express resource allocation, asynchronous effects and lifting IO computations to `F` so I think it's fine.